File: LeetCode/Medium/0036-valid-sudoku/0036-valid-sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
        int[][] row = new int[9][9];
        int[][] col = new int[9][9];
        int[][] box = new int[9][9];
        
        for (int r = 0; r < 9; ++r){
            for (int c = 0; c < 9; ++c){
                int v = board[r][c];
                
                if (v == '.') continue;

                v -= 49;

                int boxNum = (c < 3 ? 0 : c < 6 ? 1 : 2) + (3 * (r < 3 ? 0 : r < 6 ? 1 : 2));

                if (row[r][v] > 0 || col[c][v] > 0 || box[boxNum][v] > 0) return false;

                ++row[r][v];
                ++col[c][v];
                ++box[boxNum][v];
            }
        }
        
        return true;
    }
}

File: LeetCode/Medium/0036-valid-sudoku/0036-valid-sudoku.go
func isValidSudoku(board [][]byte) bool {
    row := make([][]int, 9)
    col := make([][]int, 9)
    box := make([][]int, 9)
    
    for i := 0; i < 9; i++ {
        row[i] = make([]int, 9)
        col[i] = make([]int, 9)
        box[i] = make([]int, 9)
    }
    
    for r := 0; r < 9; r++ {
        for c := 0; c < 9; c++ {
            v := int(board[r][c]) - 49
            
            if v < 0 { continue }
            
            boxNum := 0
            
            if c > 5 {
                boxNum += 2
            } else if c > 2 {
                boxNum += 1
            }
            
            if r > 5 {
                boxNum += 6
            } else if r > 2 {
                boxNum += 3
            }

            if row[r][v] > 0 || col[c][v] > 0 || box[boxNum][v] > 0 { return false }
            
            row[r][v]++
            col[c][v]++
            box[boxNum][v]++
        }
    }
    
    return true
}